core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java

```diff
@@ -54,70 +57,82 @@
+                 int defaultMaxAttempts, long responseTimeoutMillisForEachAttempt) {
+        return new RetryingRpcClientBuilder(retryStrategy)
+                .defaultMaxAttempts(defaultMaxAttempts)
+                .responseTimeoutMillisForEachAttempt(responseTimeoutMillisForEachAttempt).newDecorator();
+    }
+
+    /**
      * Creates a new instance that decorates the specified {@link Client}.
      */
     RetryingRpcClient(Client<RpcRequest, RpcResponse> delegate,
-                      RetryStrategy<RpcRequest, RpcResponse> retryStrategy, int defaultMaxAttempts) {
-        super(delegate, retryStrategy, defaultMaxAttempts);
+                      RetryStrategy<RpcRequest, RpcResponse> retryStrategy,
+                      int defaultMaxAttempts, long responseTimeoutMillisForEachAttempt) {
+        super(delegate, retryStrategy, defaultMaxAttempts, responseTimeoutMillisForEachAttempt);
     }
 
     @Override
     protected RpcResponse doExecute(ClientRequestContext ctx, RpcRequest req) throws Exception {
         final DefaultRpcResponse responseFuture = new DefaultRpcResponse();
-        retry(ctx, req, () -> {
-            try {
-                resetResponseTimeout(ctx);
-                return delegate().execute(ctx, req);
-            } catch (Exception e) {
-                return new DefaultRpcResponse(e);
-            }
-        }, responseFuture);
+        doExecute0(ctx, req, responseFuture);
         return responseFuture;
     }
 
-    private void retry(ClientRequestContext ctx, RpcRequest req, Supplier<RpcResponse> action,
-                       DefaultRpcResponse responseFuture) {
-        RpcResponse response = action.get();
+    private void doExecute0(ClientRequestContext ctx, RpcRequest req, DefaultRpcResponse responseFuture) {
+        if (!setResponseTimeout(ctx)) {
+            responseFuture.completeExceptionally(ResponseTimeoutException.get());
+            return;
+        }
+        final RpcResponse response = executeDelegate(ctx, req);
+
         retryStrategy().shouldRetry(req, response).handle(voidFunction((backoffOpt, unused) -> {
             if (backoffOpt.isPresent()) {
-                retry0(ctx, backoffOpt.get(), req, action, responseFuture);
+                long nextDelay;
+                try {
+                    nextDelay = getNextDelay(ctx, backoffOpt.get());
+                    if (nextDelay < 0) { // exceed the number of max attempt
+                        responseFuture.completeExceptionally(RetryGiveUpException.get());
+                        return;
+                    }
+                } catch (ResponseTimeoutException e) {
+                    responseFuture.completeExceptionally(e);
+                    return;
+                }
+
+                EventLoop eventLoop = ctx.contextAwareEventLoop();
+                if (nextDelay == 0) {
+                    eventLoop.submit(() -> doExecute0(ctx, req, responseFuture));
```

Comment:
If you are not interested in the returned `Future`, you can just use `execute()`

Reply:
Just copied it from the previous codes without thinking :-) Thank you~!